### PR TITLE
backend/feat: FRFS Bus Booking

### PR DESCRIPTION
**Pull request by khuzema786**
## Type of Change

- [ ] Bugfix
- [ ] New feature
- [ ] Enhancement
- [ ] Refactoring
- [ ] Dependency updates

### Flow Changes
1. Added Support for Bus Booking with any Third Party Provider.
2. Added Support for Seperate Payment Account for FRFS Bus Booking on City level.
3. Added no cancellation config based support if a provider does not allow cancellation.
4. Added support for Static Matrix based Route Pricing for each stops.

### API Level Spec Changes
**GET:** `{{baseUrl_app}}/frfs/routes?vehicleType="BUS"&city=Kochi`
Response:
```
[
    {
        "code": "Route:1:SHUTTLE-U",
        "endPoint": {
            "lat": 22.64223,
            "lon": 88.43121
        },
        "longName": "ECO T PARK-AIRPORT GATE NO1",
        "shortName": "SHUTTLEU",
        "startPoint": {
            "lat": 22.5945,
            "lon": 88.4728
        }
    }
]
```

**GET:** `{{baseUrl_app}}/frfs/stations?vehicleType="BUS"&city=Kochi&routeCode={{route_code}}`
Response:
```
[
    {
        "address": null,
        "code": "Stop:1:18024",
        "color": null,
        "distance": null,
        "lat": 22.61831,
        "lon": 88.465,
        "name": "AKANKHA MORE",
        "sequenceNum": 2,
        "stationType": "INTERMEDIATE"
    }
    ...
]
```

**POST:** `{{baseUrl_app}}/frfs/search?vehicleType="BUS"`
Request:
```
{
    "fromStationCode": "{{start_stop_code}}",
    "toStationCode": "{{end_stop_code}}",
    "routeCode": "{{route_code}}",
    "quantity": 4
}
```
Response:
```
{
    "searchId": "e3354c76-7c9c-4b2b-9c37-44ae60a2cb4b"
}
```

**GET:** `{{baseUrl_app}}/frfs/search/:searchId/quote`
Response:
```
[
    {
        "_route": {
            "code": "Route:1:SHUTTLE-U",
            "endPoint": {
                "lat": 22.64223,
                "lon": 88.43121
            },
            "longName": "ECO T PARK-AIRPORT GATE NO1",
            "shortName": "SHUTTLEU",
            "startPoint": {
                "lat": 22.5945,
                "lon": 88.4728
            }
        },
        "vehicleType": "BUS",
        ...
    }
]
```

**POST:** `{{baseUrl_app}}/frfs/quote/:quoteId/confirm`
Response:
```
{
    "_route": {
        "code": "Route:1:SHUTTLE-U",
        "endPoint": {
            "lat": 22.64223,
            "lon": 88.43121
        },
        "longName": "ECO T PARK-AIRPORT GATE NO1",
        "shortName": "SHUTTLEU",
        "startPoint": {
            "lat": 22.5945,
            "lon": 88.4728
        }
    },    
    "vehicleType": "BUS",
    ...
}
```

**GET:** `{{baseUrl_app}}/frfs/booking/:bookingId/status`
Response:
```
{
    "_route": {
        "code": "Route:1:SHUTTLE-U",
        "endPoint": {
            "lat": 22.64223,
            "lon": 88.43121
        },
        "longName": "ECO T PARK-AIRPORT GATE NO1",
        "shortName": "SHUTTLEU",
        "startPoint": {
            "lat": 22.5945,
            "lon": 88.4728
        }
    },
    "vehicleType": "BUS",
    ...
}
```

### Additional Changes

- [ ] This PR modifies the database schema (database migration added)
- [ ] This PR modifies dhall configs/environment variables
- [ ] This PR contains API breaking changes

## Checklist

- [ ] I formatted the code and addressed linter errors `./dev/format-all-files.sh`
- [ ] I reviewed submitted code
- [ ] I added unit tests for my changes where possible
- [ ] I added integration tests for my changes where possible
- [ ] I tested the changes using integration tests
- [ ] I added a [CHANGELOG](/CHANGELOG.md) entry if applicable
- [ ] No leak detected in leakcanary

### CODE REVIEW

File: Backend/app/rider-platform/rider-app/Main/src/API/Beckn/FRFS/OnConfirm.hs [hunk 1]
```diff
@@ -55,22 +50,14 @@ onConfirm _ req = withFlowHandlerAPI $ do
...
             DOnConfirm.onConfirm merchant booking onConfirmReq
-      else do
-        frfsConfig <-
-          CQFRFSConfig.findByMerchantOperatingCityId ticketBooking.merchantOperatingCityId
-            >>= fromMaybeM (InternalError $ "FRFS config not found for merchant operating city Id " <> show ticketBooking.merchantOperatingCityId)
-        void $ QFRFSTicketBooking.updateStatusById DFRFSTicketBooking.FAILED (Id bookingId)
-        void $ QFRFSTicketBookingPayment.updateStatusByTicketBookingId DFRFSTicketBookingPayment.REFUND_PENDING (Id bookingId)
-        void $ DACFOC.callBPPCancel ticketBooking bapConfig frfsConfig Spec.CONFIRM_CANCEL ticketBooking.merchantId
-
+      Nothing -> DOnConfirm.onConfirmFailure bapConfig ticketBooking
```
Comment (hkmangla) [inline Backend/app/rider-platform/rider-app/Main/src/API/Beckn/FRFS/OnConfirm.hs:60]: What if cancellation is not supported for Bus Booking?

File: Backend/app/rider-platform/rider-app/Main/src/ExternalBPP/Common/Search.hs [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+      fork "FRFS SearchReq" $ do
+        fromStation <- QStation.findById searchReq.fromStationId >>= fromMaybeM (StationNotFound searchReq.fromStationId.getId)
+        toStation <- QStation.findById searchReq.toStationId >>= fromMaybeM (StationNotFound searchReq.toStationId.getId)
+        bknSearchReq <- ACL.buildSearchReq searchReq bapConfig fromStation toStation merchantOperatingCity.city
+        logDebug $ "FRFS SearchReq " <> encodeToText bknSearchReq
+        Metrics.startMetrics Metrics.SEARCH_FRFS merchant.name searchReq.id.getId merchantOperatingCity.id.getId
+        void $ CallFRFSBPP.search bapConfig.gatewayUrl bknSearchReq merchant.id
+    (BUS, EBIX) -> do
+      onSearchReq <- EBIXFlow.search merchant merchantOperatingCity bapConfig searchReq
+      processOnSearch onSearchReq
```
Comment (hkmangla) [inline Backend/app/rider-platform/rider-app/Main/src/ExternalBPP/Common/Search.hs:32]: Put it in `fork`

File: Backend/app/rider-platform/rider-app/Main/src/ExternalBPP/Common/Init.hs [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+  case (bapConfig.vehicleCategory, bapConfig.provider) of
+    (METRO, ONDC) -> do
+      providerUrl <- booking.bppSubscriberUrl & parseBaseUrl & fromMaybeM (InvalidRequest "Invalid provider url")
+      bknInitReq <- ACL.buildInitReq (mRiderName, mRiderNumber) booking bapConfig Utils.BppData {bppId = booking.bppSubscriberId, bppUri = booking.bppSubscriberUrl} merchantOperatingCity.city
+      logDebug $ "FRFS InitReq " <> encodeToText bknInitReq
+      Metrics.startMetrics Metrics.INIT_FRFS merchant.name booking.searchId.getId merchantOperatingCity.id.getId
+      void $ CallFRFSBPP.init providerUrl bknInitReq merchant.id
+    (BUS, EBIX) -> do
+      onInitReq <- EBIXFlow.init merchant merchantOperatingCity bapConfig (mRiderName, mRiderNumber) booking
+      processOnInit onInitReq
```
Comment (hkmangla) [inline Backend/app/rider-platform/rider-app/Main/src/ExternalBPP/Common/Init.hs:30]: fork

**Review by hkmangla — CHANGES_REQUESTED** [review]

File: Backend/app/rider-platform/rider-app/Main/src/API/Beckn/FRFS/OnConfirm.hs (hunk 1 above)
Comment (khuzema786) [reply Backend/app/rider-platform/rider-app/Main/src/API/Beckn/FRFS/OnConfirm.hs:59]: Throwing CancellationNotSupported Error in the APIs that can cancel the Booking.
```
endpoint: /frfs/booking/{bookingId}/canCancel
endpoint: /frfs/booking/{bookingId}/cancel
```

**Review by hkmangla — APPROVED** [review]

**Review by hkmangla — APPROVED** [review]

**Pull request merged**